include/LightGBM/c_api.h

```diff
@@ -10,55 +18,45 @@
 #include <cstdint>
 #include <cstring>
 
-/*!
-* To avoid type conversion on large data, most of our expose interface support both for float_32 and float_64.
-* Except following:
-* 1. gradients and hessians.
-* 2. Get current score for training data and validation
-* The reason is because they are called frequently, the type-conversion on them maybe time cost.
-*/
 
-typedef void* DatasetHandle;
-typedef void* BoosterHandle;
+typedef void* DatasetHandle;  /*!< \brief Handle of dataset. */
+typedef void* BoosterHandle;  /*!< \brief Handle of booster. */
 
-#define C_API_DTYPE_FLOAT32 (0)
-#define C_API_DTYPE_FLOAT64 (1)
-#define C_API_DTYPE_INT32   (2)
-#define C_API_DTYPE_INT64   (3)
-#define C_API_DTYPE_INT8    (4)
+#define C_API_DTYPE_FLOAT32 (0)  /*!< \brief float32 (single precision float). */
+#define C_API_DTYPE_FLOAT64 (1)  /*!< \brief float64 (double precision float). */
+#define C_API_DTYPE_INT32   (2)  /*!< \brief int32. */
+#define C_API_DTYPE_INT64   (3)  /*!< \brief int64. */
+#define C_API_DTYPE_INT8    (4)  /*!< \brief int8. */
 
-#define C_API_PREDICT_NORMAL     (0)
-#define C_API_PREDICT_RAW_SCORE  (1)
-#define C_API_PREDICT_LEAF_INDEX (2)
-#define C_API_PREDICT_CONTRIB    (3)
+#define C_API_PREDICT_NORMAL     (0)  /*!< \brief Normal prediction, with transform (if needed). */
+#define C_API_PREDICT_RAW_SCORE  (1)  /*!< \brief Predict raw score. */
+#define C_API_PREDICT_LEAF_INDEX (2)  /*!< \brief Predict leaf index. */
+#define C_API_PREDICT_CONTRIB    (3)  /*!< \brief Predict feature contributions (SHAP values). */
 
 /*!
- * \fn LGBM_GetLastError
- * \headerfile <LightGBM/export.h>
  * \brief Get string message of the last error.
- * \return error information
-*/
+ * \return Error information
+ */
 LIGHTGBM_C_EXPORT const char* LGBM_GetLastError();
 
 // --- start Dataset interface
 
 /*!
- * \fn LGBM_DatasetCreateFromFile
  * \brief Load dataset from file (like LightGBM CLI version does).
  * \param filename The name of the file
  * \param parameters Additional parameters
  * \param reference Used to align bin mapper with other dataset, nullptr means isn't used
  * \param[out] out A loaded dataset
  * \return 0 when succeed, -1 when failure happens
-*/
+ */
 LIGHTGBM_C_EXPORT int LGBM_DatasetCreateFromFile(const char* filename,
                                                  const char* parameters,
                                                  const DatasetHandle reference,
                                                  DatasetHandle* out);
 
 /*!
- * \fn LGBM_DatasetCreateFromSampledColumn
  * \brief Create an empty dataset by sampling data.
+ * \param sample_data Sampled data, grouped by the column
```

Comment:
This lost param was the reason why #2076 ruined all C API docs.
It's quite strange, but `\fn` directive makes Doxygen produce xml without descriptions instead of generating the error.
Also, Doxygen's docs say that we don't need to use `\fn` directives when comments are placed directly before function:
> Warning
Do not use this command if it is not absolutely needed, since it will lead to duplication of information and thus to errors.
http://www.doxygen.nl/manual/commands.html#cmdfn

So I removed them all.